### PR TITLE
Column equality testing fixes

### PR DESCRIPTION
**Pull request by brandon-b-miller**
Fixes a bug where empty columns were not comparing correctly as well as a few edge cases with strings

Partially addresses https://github.com/rapidsai/cudf/issues/8513

### CODE REVIEW

File: python/cudf/cudf/testing/testing.py [hunk 1]
```diff
@@ -201,31 +206,50 @@ def assert_column_equal(
...
+        except TypeError as e:
+            if str(e) != "Categoricals can only compare with the same type":
+                raise e
+            if is_categorical_dtype(left) and is_categorical_dtype(right):
+                left = left.astype(left.categories.dtype)
+                right = right.astype(right.categories.dtype)
     if not columns_equal:
-        msg1 = f"{left.values_host}"
-        msg2 = f"{right.values_host}"
+        ldata = [val for val in left.to_pandas(nullable=True)]
```
Comment (isVoid) [inline python/cudf/cudf/testing/testing.py:244]: Side question - I've wondered how reliable `nullable=True` is given pandas support for nullable float dtype is still non-public?

**Review by isVoid — COMMENTED** [review]
Maybe we should add some tests in `test_testing.py` to cover these changes.

File: python/cudf/cudf/testing/testing.py [hunk 2]
```diff
@@ -201,31 +206,50 @@ def assert_column_equal(
...
-                cp.all(left.isnull().values == right.isnull().values)
-                and cp.allclose(
-                    left[left.isnull().unary_operator("not")].values,
-                    right[right.isnull().unary_operator("not")].values,
+    if left.size == right.size == 0:
+        columns_equal = True
+    elif not (
+        (is_string_dtype(left) and is_numeric_dtype(right))
+        or (is_numeric_dtype(left) and is_string_dtype(right))
+    ):
```
[the remark this answers is not in the document]
Comment (vyasr) [reply python/cudf/cudf/testing/testing.py:215]: Yeah this seems odd. Are we just generally looking to avoid checking this for mismatched dtypes? And doesn't cupy not support strings anyway, so if left and right are both strings we'll go into this block and then fail (the exception handling only deals with categoricals)?

File: python/cudf/cudf/testing/testing.py [hunk 3]
```diff
@@ -201,31 +206,50 @@ def assert_column_equal(
...
+                        cp.all(left.isnull().values == right.isnull().values)
+                        if not (left.dtype.kind == right.dtype.kind == "f")
+                        else (
+                            cp.all(is_nan(left).values == is_nan(right).values)
+                            and cp.all(
+                                left.isnull().values == right.isnull().values
+                            )
+                        )
+                    )
+                    and cp.allclose(
```
[the remark this answers is not in the document]
Comment (brandon-b-miller) [reply python/cudf/cudf/testing/testing.py]: I didn't see a way of using `equal_nan` but I agree that the logic was a little hard to follow in general, so I redid it here. Let me know if you think this is better.

File: python/cudf/cudf/testing/testing.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (brandon-b-miller) [reply python/cudf/cudf/testing/testing.py:215]: If one column is string and the other is numeric, we can assume the columns are not equal. (`1 != '1')`. These lines check to make sure exactly one of the columns is string and the other is numeric, in which case we avoid the entire try/except block and therefore avoid any opportunity for `columns_equal` to be set to `True`.  We should end up on line 243 from there.

File: python/cudf/cudf/testing/testing.py (hunk 1 above)
Comment (brandon-b-miller) [reply python/cudf/cudf/testing/testing.py:243]: It seems like they're being fairly forward with these [as of 1.2.0](https://pandas.pydata.org/pandas-docs/stable/whatsnew/v1.2.0.html?highlight=float64dtype#experimental-nullable-data-types-for-float-data)

**Comment by brandon-b-miller** [comment]
everything look ok here? cc @vyasr @isVoid

File: python/cudf/cudf/testing/testing.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (vyasr) [reply python/cudf/cudf/testing/testing.py:215]: Is this only an issue for string types, or do we need to worry about other types as well? I see categoricals are handled above, what about list/struct dtypes?

File: python/cudf/cudf/testing/testing.py [hunk 4]
```diff
@@ -201,31 +206,49 @@ def assert_column_equal(
...
-                and cp.allclose(
+    if left.size == right.size == 0:
+        columns_equal = True
+    elif not (
+        (is_string_dtype(left) and is_numeric_dtype(right))
+        or (is_numeric_dtype(left) and is_string_dtype(right))
+    ):
+        try:
+            # nulls must be in the same places for all dtypes
+            nulls_equal = cp.all(left.isnull().values == right.isnull().values)
```
Comment (vyasr) [inline python/cudf/cudf/testing/testing.py:218]: Do you need to access the values here? I think you can just rely on `__cuda_array_interface__`.

File: python/cudf/cudf/testing/testing.py [hunk 5]
```diff
@@ -201,31 +206,49 @@ def assert_column_equal(
...
+                    )
+                    columns_equal = nulls_equal and values_equal and nans_equal
+            else:
+                columns_equal = left.equals(right)
+        except TypeError as e:
+            if str(e) != "Categoricals can only compare with the same type":
+                raise e
+            if is_categorical_dtype(left) and is_categorical_dtype(right):
+                left = left.astype(left.categories.dtype)
+                right = right.astype(right.categories.dtype)
```
Comment (vyasr) [inline python/cudf/cudf/testing/testing.py:241]: These if elses become a little cleaner if you eschew the `nulls_equal` and `nans_equal` variables in favor of just assigning back to `columns_equal`. I can't suggest because there's intervening unchanged lines, but basically just do
```
columns_equal = cp.all(...) # Replace assigning to nulls_equal
if columns_equal and not check_exact...:
    columns_equal = cp.allclose...
    if columns_equal and (left.dtype.kind == right.dtype.kind == "f"): # Note that I've inverted the second condition
        columns_equal = cp.all(...)
```
This also removes unnecessary checks if one of the previous ones has already failed.

File: python/cudf/cudf/testing/testing.py [hunk 6]
```diff
@@ -201,31 +206,49 @@ def assert_column_equal(
...
     if not columns_equal:
-        msg1 = f"{left.values_host}"
-        msg2 = f"{right.values_host}"
+        ldata = [val for val in left.to_pandas(nullable=True)]
+        rdata = [val for val in right.to_pandas(nullable=True)]
+        msg1 = f"{ldata}"
+        msg2 = f"{rdata}"
```
Comment (vyasr) [inline python/cudf/cudf/testing/testing.py:246]: These can just be inlined into the `raise_assert_detail` function call. Pass by kwarg if you think that's helpful in documenting the purpose of these messages.

File: python/cudf/cudf/testing/testing.py [hunk 7]
```diff
@@ -234,6 +257,12 @@ def assert_column_equal(
         )
 
 
+def null_safe_scalar_equals(left, right):
+    if left in {cudf.NA, np.nan} or right in {cudf.NA, np.nan}:
+        return left is right
+    return left == right
```
Comment (vyasr) [inline python/cudf/cudf/testing/testing.py:263]: Could we just do 
```suggestion
    return (left is right) or (left == right)
```
?

**Review by vyasr — CHANGES_REQUESTED** [review]
Some suggestions for improvement here and one question.

File: python/cudf/cudf/testing/testing.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (brandon-b-miller) [reply python/cudf/cudf/testing/testing.py:215]: I looked into what happens here and it would appear that when `list` or `struct` are involved we fall into the `left.equals(right)` check at the very end and end up with a `TypeError`, except in the case that we're comparing `list` to `list` or `struct` to `struct`. That should probably not happen.

I see the point I think you are making though: there's a certain set of dtypes (beyond just string) where, even if `check_dtype=False`, we know up front that it's a 100% mismatch between non-null elements simply because a struct can't compare as equal to a "not struct". I think these dtypes are:

- String
- List
- Struct
- Interval
- Decimal

I suppose the only edge case here would be that we should arguably return `True` even for that subset of dtypes if the columns are fully null. 

If this seems like the correct logic I am happy to go back and wire it up as such here.

[the remark this answers is not in the document]
Comment (vyasr) [reply python/cudf/cudf/testing/testing.py:215]: Yep that all sounds good. For the fully null case, I think whether or not we return True should be determined by `check_dtypes`, but if someone has a different opinion I'm open to a different result.

File: python/cudf/cudf/testing/testing.py (hunk 7 above)
Comment (brandon-b-miller) [reply python/cudf/cudf/testing/testing.py:293]: The problem with this approach is that if `left == NaN` and `right == NA`, the first check evaluates false and the second one evaluates to `NA`, making the whole expression evaluate to `NA`. This ends up creating an `if NA:` statement from the output and triggers a TypeError. This causes the `diff` to read 100% and we get the wrong error message. The original function avoids this situation by never allowing `NA` or `NaN` to pass to the `==` check. I added a test

File: python/cudf/cudf/testing/testing.py (hunk 5 above)
Comment (brandon-b-miller) [reply python/cudf/cudf/testing/testing.py]: This worked great, but also requires a quick update to the categorical handling below.

File: python/cudf/cudf/testing/testing.py (hunk 4 above)
Comment (brandon-b-miller) [reply python/cudf/cudf/testing/testing.py]: It looks like `cp.all` only plays nice with `cupy` arrays:

```python
>>> cupy.all(
...     cudf.core.column.as_column([True, False, True])
... )
TypeError: 'a' must be a cupy.ndarray object, not <class 'cudf.core.column.numerical.NumericalColumn'>.
```

Comment (vyasr) [reply python/cudf/cudf/testing/testing.py]: I see, that's kind of annoying that it doesn't try to convert internally.

File: python/cudf/cudf/testing/testing.py [hunk 8]
```diff
@@ -210,38 +229,46 @@ def assert_column_equal(
...
             if is_categorical_dtype(left) and is_categorical_dtype(right):
                 left = left.astype(left.categories.dtype)
                 right = right.astype(right.categories.dtype)
     if not columns_equal:
-        ldata = [val for val in left.to_pandas(nullable=True)]
-        rdata = [val for val in right.to_pandas(nullable=True)]
+        ldata = str([val for val in left.to_pandas(nullable=True)])
+        rdata = str([val for val in right.to_pandas(nullable=True)])
         msg1 = f"{ldata}"
         msg2 = f"{rdata}"
```
Comment (vyasr) [inline python/cudf/cudf/testing/testing.py:273]: You can delete these now.

File: python/cudf/cudf/testing/testing.py (hunk 7 above)
Comment (vyasr) [reply python/cudf/cudf/testing/testing.py:293]: Gotcha, yeah that makes sense.

**Review by vyasr — APPROVED** [review]
One last suggestion, otherwise good on my end

**Review by isVoid — APPROVED** [review]
It's all good. Just minor stuff below.

**Comment by brandon-b-miller** [comment]
@gpucibot merge
